src/block-cache.ts

```diff
@@ -0,0 +1,183 @@
+  }
+
+  // removes all block caches with block number lower than `oldBlockHex`
+  clearBefore(oldBlockHex: string): void {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const self = this;
+    const oldBlockNumber: number = Number.parseInt(oldBlockHex, 16);
+    // clear old caches
+    Object.keys(self.cache)
+      .map(Number)
+      .filter((num) => num < oldBlockNumber)
+      .forEach((num) => delete self.cache[num]);
+  }
+}
+
+function createBlockCacheMiddleware(
+  opts: BlockCacheMiddlewareOptions = {},
+): JsonRpcMiddleware<string[], Record<string, unknown>> {
+  // validate options
+  const { blockTracker } = opts;
+
+  if (!blockTracker) {
+    throw new Error('createBlockCacheMiddleware - No BlockTracker specified');
+  }
+
+  // create caching strategies
+  const blockCache: BlockCacheStrategy = new BlockCacheStrategy();
+  const strategies: Record<string, BlockCacheStrategy> = {
+    perma: blockCache,
+    block: blockCache,
+    fork: blockCache,
+  };
+
+  return createAsyncMiddleware(async (req, res, next) => {
+    // allow cach to be skipped if so specified
+    if (((req as unknown) as Record<string, unknown>).skipCache) {
+      return next();
+    }
+    // check type and matching strategy
+    const type: string = cacheUtils.cacheTypeForPayload(req);
+    const strategy: BlockCacheStrategy = strategies[type];
+    // If there's no strategy in place, pass it down the chain.
+    if (!strategy) {
+      return next();
+    }
+    // If the strategy can't cache this request, ignore it.
+    if (!strategy.canCacheRequest(req)) {
+      return next();
+    }
+
+    // get block reference (number or keyword)
+    let blockTag: string|null = cacheUtils.blockTagForPayload(req);
+    if (!blockTag) {
+      blockTag = 'latest';
+    }
+
+    // get exact block number
+    let requestedBlockNumber: string;
+    if (blockTag === 'earliest') {
+      // this just exists for symmetry with "latest"
+      requestedBlockNumber = '0x00';
+    } else if (blockTag === 'latest') {
+      // fetch latest block number
+      const latestBlockNumber: string = await blockTracker.getLatestBlock();
+      // clear all cache before latest block
+      blockCache.clearBefore(latestBlockNumber);
+      requestedBlockNumber = latestBlockNumber;
+    } else {
+      // We have a hex number
+      requestedBlockNumber = blockTag;
+    }
+
+    // end on a hit, continue on a miss
+    const cacheResult: Record<string, unknown> = await strategy.get(req, requestedBlockNumber) as Record<string, unknown>;
+    if (cacheResult === undefined) {
+      // cache miss
+      // wait for other middleware to handle request
+      // eslint-disable-next-line node/callback-return
+      await next();
+      // add result to cache
+      await strategy.set(req, requestedBlockNumber, (res.result as Record<string, unknown>));
+    } else {
+      // fill in result from cache
+      res.result = (cacheResult as Record<string, unknown>);
```

Comment:
Per feedback elsewhere in this function, it would be great if we could do less `Record<string, unknown>` and casting here. Probably (also as suggested elsewhere), by manually defining some interfaces.